funcbench/main.go

```diff
@@ -0,0 +1,281 @@
+		log.Println(string(data))
+		if err != nil {
+			log.Fatalln(err)
+		}
+		if err := os.Chdir(repo); err != nil {
+			log.Fatalln(err)
+		}
+
+		latestCommitHash := os.Getenv("GITHUB_SHA")
+		logLink := fmt.Sprintf("Check https://github.com/%s/%s/commit/%s/checks for more information.", owner, repo, latestCommitHash)
+
+		data, err = exec.Command("git", "config", "--global", "user.email", "prombench@example.com").CombinedOutput()
+		if err != nil {
+			log.Fatalln("Error: ", err, "Command out: ", string(data))
+		}
+		data, err = exec.Command("git", "config", "--global", "user.name", "Prombench Bot Junior").CombinedOutput()
+		if err != nil {
+			log.Fatalln("Error: ", err, "Command out: ", string(data))
+		}
+
+		// Branch setup.
+		data, err = exec.Command("git", "fetch", "origin", fmt.Sprintf("pull/%d/head:pullrequest", prNumber)).CombinedOutput()
+		if err != nil {
+			log.Fatalln("Error: ", err, "Command out: ", string(data))
+		}
+
+		data, err = exec.Command("git", "checkout", branch).CombinedOutput()
+		if err != nil {
+			log.Fatalln("Error: ", err, "Command out: ", string(data))
+		}
+
+		cmd := exec.Command("git", "merge", "--squash", "--no-commit", "pullrequest")
+		data, err = cmd.CombinedOutput()
+		if err != nil {
+			log.Fatalln("Error: ", err, "Command out: ", string(data))
+		}
+		if cmd.ProcessState.ExitCode() != 0 {
+			if err := postComment(clt, owner, repo, prNumber, "Git merge failed"); err != nil {
+				log.Fatalln(err)
+			}
+		}
+
+		data, err = exec.Command("git", "reset").CombinedOutput()
+		if err != nil {
+			log.Fatalln("Error: ", err, "Command out: ", string(data))
+		}
+
+		// Benchmark the with pullrequest changes.
+		if race == "-no-race" {
+			cmd = exec.Command("go", "test", "-bench", fmt.Sprintf("^%s$", issueComment), "-benchmem", "-v", "./...")
+		} else {
+			cmd = exec.Command("go", "test", "-bench", fmt.Sprintf("^%s$", issueComment), "-benchmem", "-race", "-v", "./...")
+		}
+		data, err = cmd.CombinedOutput()
+		log.Println(string(data))
+		if err != nil || cmd.ProcessState.ExitCode() != 0 {
+			if err := postComment(clt, owner, repo, prNumber, fmt.Sprintf("Go test for this pull request failed. %s", logLink)); err != nil {
+				log.Fatalln(err)
+			}
+			log.Fatalln(err)
+		}
+		err = ioutil.WriteFile(newPath, data, 0644)
+		if err != nil {
+			log.Fatalln(err)
+		}
+
+		// Checkout to the comparing branch.
+		err = filepath.Walk(os.Getenv("GITHUB_WORKSPACE"), func(path string, info os.FileInfo, err error) error {
+			if err != nil {
+				return err
+			}
+			if info.IsDir() || strings.Contains(path, ".git/") || strings.HasSuffix(info.Name(), "test.go") {
+				return nil
+			}
+
+			data, err = exec.Command("git", "checkout", "--", path).CombinedOutput()
+			if err != nil {
+				log.Println(string(data))
+
+				// New file appeared and is not tracked, then it can be removed
+				if strings.Contains(string(data), "did not match any file") {
+					data, err = exec.Command("rm", path).CombinedOutput()
```

Comment:
When would such file appear and why do you have to remove it?

Reply:
When I started testing Prometheus, it created a file called `lock`. When the program executed `git checkout -- lock`, an error occurred because `lock` is not tracked, and therefore, cannot be checked out. This lead to a failure.

Reply:
ok thanks.